### ISSUE TITLE
Ability To Customize File Path

### ISSUE THREAD

**Issue by johnmasonwilliams** [ryanto/next-s3-upload#61]
Hello! I just found and implemented this library into my project but would like a way to customize the path of my files.

A simple path parameter that we could pass alongside our file in `uploadToS3` would be fantastic.

If there is already a way to do this, adding that to the docs would be super helpful. Thanks!

**Comment by johnmasonwilliams** [comment]
To add on, I understand you can configure the file path in the api route but I was more thinking of simply an optional parameter in the `uploadToS3` function for a path.

**Comment by ryanto** [comment]
Thanks for opening an issue!

Just to clarify, you want to do key customization, but have the key come from the client side app?

**Comment by jchaing** [comment]
> Thanks for opening an issue!
> 
> Just to clarify, you want to do key customization, but have the key come from the client side app?

I would so buy you a beer if we could do that. We're trying to upload images to sub-folders in our S3 Bucket dynamically by user.id. We could achieve this by adding it to a query param in the URL and extract it from req.headers.referer, but didn't want to expose the id in the URL.

**Comment by johnmasonwilliams** [comment]
Yep, that's exactly right. I have a digital marketplace and structure my s3 bucket like 'Products/<PRODUCT_ID>/Images/<FILE_NAME>' for product images or 'Products/<PRODUCT_ID>/Files/<FILE_NAME>' for product deliverables.

**Comment by ryanto** [comment]
Ok cool. I absolutely think we can get this added.

What do you think about this API: 

```jsx
// client component

let { uploadToS3 } = useS3Upload();

function handleUpload(file) {
  let upload = await uploadToS3(file, { 
    endpoint: {
      request: {
        body: { 
          isImage: true,
          productId: "abc123"
        } 
     } 
  });
}
```

This would modify the body that goes to your next api route, so you could access it in the key function like so.

```js
// on the server

export default APIRoute.configure({
  key(req, filename) {
    return `Products/${req.body.productId}/${req.body.isImage ? 'Images' : 'Files'}/${filename}`;
  },
});
```

**Comment by johnmasonwilliams** [comment]
That does work for my use case! I assume that the body could be composed of anything though correct? Like instead of `isImage` and `productId` it could be just `path`?

Something like this?

```
// client component

let { uploadToS3 } = useS3Upload();

function handleUpload(file) {
  let upload = await uploadToS3(file, { 
    endpoint: {
      request: {
        body: { 
          path: `Products/1234/Images/`
        } 
     } 
  });
}
```

```
// on the server

export default APIRoute.configure({
  key(req, filename) {
    return `${req.body.path}/${filename}`;
  },
});
```

If so, then that would work great for my use case and I'm sure many other use cases!

**Comment by johnmasonwilliams** [comment]
If we were able to simplify it down even more to just adding another parameter for `path` that would be great. If it is too much than the current implementation works great.

Something like this:
```
// client component

let { uploadToS3 } = useS3Upload();

function handleUpload(file) {
  let upload = await uploadToS3(file, {
    path: `Products/1234/Images/`
  });
}

// or

function handleUpload(file) {
  let path = 'Products/1234/Images/'
  let upload = await uploadToS3(file, path)
}
```

```
// on the server

export default APIRoute.configure({
  key(req, filename, path) {
    return `${path}/${filename}`;
  },
});
```

**Comment by ryanto** [comment]
Awesome! Glad to hear it will work!

One issue with explicitly coding a `path`  property is that it can only be used for the path. On one hand, it's nice that its specific and constrained, but on the other hand if someone comes along asking to customize another part of S3 upload then we're back to the drawing board.

With `endpoint.request` we offer a low level API that can potentially be used for other use cases.

Since next-s3-upload has to communicate with the api route to get a signature we can piggy back off that request. In this case we're passing along post body data, but you could imagine in the future we let people add something like headers here. We don't have to come up with a new api for that, since we'd have this low level idea of modifying the request already built into the library.

```js
function handleUpload(file) {
  let upload = await uploadToS3(file, { 
    endpoint: {
      request: {
        body: { 
          isImage: true,
          productId: "abc123"
        } 
     } 
  });
}
```

One way I like to think about this is that this is a sharp API for power users. It's sharp because it's not as explicitly helpful as a new `path` property, but since it's made for power users maybe that's okay. Most folks just want to uploadToS3, but a power user might want to do all sorts of customizations. Let's let them modify the http request that's sent to our API route and see what sort of things they start doing.

**Comment by ryanto** [comment]
> Like instead of isImage and productId it could be just path?

```js
request: {
  body: { 
    path: 'Products/1234/Images/'
  } 
} 
```

Yes you can absolutely do this! One issue might be that you are letting the client come up with this key, so a browser could potentially do something like....

```js
request: {
  body: { 
    path: 'bet/you/werent/expecting/me/to/upload/here/'
  } 
} 
```

That said, sometimes you know that none of your users are going to send requests like that. So if it's easier to use `body.path` then go for it!

**Comment by johnmasonwilliams** [comment]
Ok, that is all super useful information. I think this would be a great addition to this next-s3-upload and like you said how this would give additional ability to power users like myself.

I really appreciate the responsiveness!

**Comment by ryanto** [comment]
Ok cool. It'll probably take me a few weeks to get this implemented just because I'm a little slammed with work. If you want to take on this feature I can help get you setup with a testing environment and guide you through exactly what needs to be done. If you're busy or not interested in doing this that's totally okay.

**Comment by johnmasonwilliams** [comment]
Unfortunately I don't have time to work on it either. For now I will just use it as is until the feature is pushed.

I appreciate your time working on this!

**Comment by Zverizsume** [comment]
@ryanto 

Hi, I just started using the lib. Can you please add some of this stuff mentioned here to the official documentation on the website. Thank you in advance!

**Comment by ryanto** [comment]
Hi @Zverizsume 

The stuff mentioned here is us discussing future code, none of this exists in the library today. It'll be added to the docs when it's ready.

**Comment by ryanto** [comment]
Ok this is all set, it's released in version `0.2.1`!

The request from the frontend can pass extra data to the key function. Docs are here: https://next-s3-upload.codingvalue.com/s3-file-paths#data-from-react

Just a heads up: There were some breaking changes in `0.2.0`, so if you're upgrading from an older version make sure you check out the [release notes on 0.2.0](https://github.com/ryanto/next-s3-upload/releases/tag/v0.2.0)


If anyone runs into any trouble let me know.
